pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,386 @@
+		return fmt.Errorf("terminating nodeclaims, %w", multiErr)
+	}
+	return nil
+}
+
+// Add will launch replacement nodeClaims and add the command to the queue
+// Each command added to the queue should already be validated and ready for execution.
+func (q *Queue) Add(ctx context.Context, candidates []*state.StateNode, replacements []*scheduling.NodeClaim, reason string) error {
+	providerIDs := lo.Map(candidates, func(s *state.StateNode, _ int) string {
+		return s.ProviderID()
+	})
+	// Cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := q.RequireNoScheduleTaint(ctx, true, candidates...); err != nil {
+		return multierr.Append(fmt.Errorf("cordoning nodes, %w", err), q.RequireNoScheduleTaint(ctx, false, candidates...))
+	}
+
+	var nodeClaimKeys []nodeclaim.Key
+	var err error
+	if len(replacements) > 0 {
+		if nodeClaimKeys, err = q.launchReplacementNodeClaims(ctx, replacements, reason); err != nil {
+			// If we failed to launch the replacement, don't disrupt.  If this is some permanent failure,
+			// we don't want to disrupt workloads with no way to provision new nodes for them.
+			return multierr.Append(fmt.Errorf("launching replacement nodeclaim, %w", err), q.RequireNoScheduleTaint(ctx, false, candidates...))
+		}
+	}
+
+	// We have the new nodeclaims created at the API server so mark the old nodeclaims for deletion
+	q.cluster.MarkForDeletion(providerIDs...)
+	// Wait for a heuristic period of time to handle initial eventual consistency delay with node claim creation.
+	// This allows us to properly exit when we detect not found errors caused by ICE errors or the Node initializationTTL.
+	cmd := NewCommand(nodeClaimKeys, candidates, reason, q.clock.Now().Add(5*time.Second))
+	q.RateLimitingInterface.AddAfter(cmd, 5*time.Second)
+	for _, candidate := range candidates {
+		q.ProviderIDToCommand[candidate.ProviderID()] = cmd
+	}
+	return nil
+}
+
+// CanAdd is a quick check to see if the candidate is already part of a disruption action
+func (q *Queue) CanAdd(ids ...string) error {
+	var err error
+	for _, id := range ids {
+		if _, ok := q.ProviderIDToCommand[id]; ok {
+			err = multierr.Append(err, fmt.Errorf("candidate is being disrupted"))
+		}
+	}
+	return err
+}
+
+// Remove fully clears the queue of all references of a hash/command
+func (q *Queue) Remove(cmd *Command) {
+	// Remove all candidates linked to the command
+	for _, candidate := range cmd.Candidates {
+		delete(q.ProviderIDToCommand, candidate.ProviderID())
+	}
+	q.RateLimitingInterface.Forget(cmd)
+	q.RateLimitingInterface.Done(cmd)
+}
+
+// RequireNoScheduleTaint will add/remove the karpenter.sh/disruption:NoSchedule taint from the candidates.
+// This is used to enforce no taints at the beginning of disruption, and
+// to add/remove taints while executing a disruption action.
+// nolint:gocyclo
+func (q *Queue) RequireNoScheduleTaint(ctx context.Context, addTaint bool, nodes ...*state.StateNode) error {
```

Comment:
Since we need this in multiple places, is it worth making this more of a helper function than a queue method?